brave-grpc/src/main/java/com/github/kristofa/brave/grpc/BraveGrpcClientInterceptor.java

```diff
@@ -0,0 +1,128 @@
+import com.github.kristofa.brave.ClientResponseAdapter;
+import com.github.kristofa.brave.ClientResponseInterceptor;
+import com.github.kristofa.brave.ClientSpanThreadBinder;
+import com.github.kristofa.brave.IdConversion;
+import com.github.kristofa.brave.KeyValueAnnotation;
+import com.github.kristofa.brave.internal.Nullable;
+import com.google.common.base.Strings;
+import com.twitter.zipkin.gen.Span;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.Status.Code;
+
+import java.util.Collection;
+import java.util.Collections;
+
+public class BraveGrpcClientInterceptor implements ClientInterceptor {
+
+    private String clientService;
+    private ClientRequestInterceptor clientRequestInterceptor;
+    private ClientResponseInterceptor clientResponseInterceptor;
+    private ClientSpanThreadBinder clientSpanThreadBinder;
+
+    public BraveGrpcClientInterceptor(String clientService, Brave brave) {
+        if (Strings.isNullOrEmpty(clientService)) {
+            throw new IllegalStateException("clientService name cannot be empty");
+        }
+        this.clientService = clientService;
+        this.clientRequestInterceptor = checkNotNull(brave.clientRequestInterceptor());
+        this.clientResponseInterceptor = checkNotNull(brave.clientResponseInterceptor());
+        this.clientSpanThreadBinder = checkNotNull(brave.clientSpanThreadBinder());
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(final MethodDescriptor<ReqT, RespT> method,
+                                                               final CallOptions callOptions, final Channel next) {
+        return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+
+            @Override
+            public void start(Listener<RespT> responseListener, Metadata headers) {
+                clientRequestInterceptor.handle(new GrpcClientRequestAdapter<>(method, headers));
+                final Span currentClientSpan = clientSpanThreadBinder.getCurrentClientSpan();
+                super.start(new SimpleForwardingClientCallListener<RespT>(responseListener) {
+                    @Override
+                    public void onClose(Status status, Metadata trailers) {
+                        clientSpanThreadBinder.setCurrentSpan(currentClientSpan);
+                        clientResponseInterceptor.handle(new GrpcClientResponseAdapter(status));
+                        super.onClose(status, trailers);
+                   }
+                }, headers);
+            }
+        };
+
+    }
+
+    private class GrpcClientRequestAdapter<ReqT, RespT> implements ClientRequestAdapter {
```

Comment:
static final class is better, I think. removing private prevents the compiler from having to [generate synthetic methods](http://www.javaworld.com/article/2073578/java-s-synthetic-methods.html). static because AFAICT, there's no shared state with the outer type.